src/inmanta/module.py

```diff
@@ -2958,22 +2957,30 @@ def get_plugin_dir(self) -> Optional[str]:
         raise NotImplementedError()
 
     def _list_python_files(self, plugin_dir: str) -> list[str]:
-        """
-        Generate a list of all Python files in the given plugin directory.
-        This method prioritizes .pyc files over .py files, uses caching to avoid duplicate directory walks,
-        and only considers directories that are Python packages.
-        """
+        """Generate a list of all python files, including namespace packages and excluding the model directory."""
         # Return cached results if this directory has been processed before
         if plugin_dir in self._dir_cache:
             return self._dir_cache[plugin_dir]
 
         files: dict[str, str] = {}
+        model_dir_path: str = os.path.join(plugin_dir, "inmanta_plugins", self.name, "model")
 
-        for dirpath, dirnames, filenames in os.walk(plugin_dir):
-            # Skip non-package directories (those without an __init__.py or __init__.pyc file)
-            if not any(fname for fname in filenames if fname in ["__init__.py", "__init__.pyc"]):
-                continue
+        def exclude_dir_walk(top, exclude_dirs: list[str]):
+            """
+            An os.walk that can stop descending into specific directories.
+
+            :param top: The root directory from which to start walking.
+            :param exclude_dirs: A list of paths of directories to stop descending into.
+            """
+            for dirpath, dirnames, filenames in os.walk(top, topdown=True):
+                # Check if the current directory matches any path in exclude_dirs
+                if any(dirpath.startswith(stop_dir) for stop_dir in exclude_dirs):
+                    # Modify dirnames in-place to stop os.walk from descending into any more subdirectories
+                    dirnames[:] = []
+                    continue
+                yield dirpath, dirnames, filenames
```

Comment:
As this is on the critical path, I would avoid the double yield  and not wrap this in an extra method.